api/docgen/openrpc.go

```diff
@@ -0,0 +1,225 @@
+// Package docgen generates an OpenRPC spec for the Celestia Node. It has been inspired by and adapted from Filecoin's
+// Lotus API implementation.
+package docgen
+
+import (
+	"encoding/json"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"net"
+	"reflect"
+	"strings"
+
+	"github.com/alecthomas/jsonschema"
+	go_openrpc_reflect "github.com/etclabscore/go-openrpc-reflect"
+	meta_schema "github.com/open-rpc/meta-schema"
+)
+
+type Visitor struct {
+	Methods map[string]ast.Node
+}
+
+func (v *Visitor) Visit(node ast.Node) ast.Visitor {
+	st, ok := node.(*ast.TypeSpec)
+	if !ok {
+		return v
+	}
+
+	if st.Name.Name != "Module" {
+		return nil
+	}
+
+	iface := st.Type.(*ast.InterfaceType)
+	for _, m := range iface.Methods.List {
+		if len(m.Names) > 0 {
+			v.Methods[m.Names[0].Name] = m
+		}
+	}
+
+	return v
+}
+
+type Comments = map[string]string
+
+func ParseCommentsFromNodebuilderModules(moduleNames ...string) Comments {
+	fset := token.NewFileSet()
+	nodeComments := make(Comments)
+	for _, moduleName := range moduleNames {
+		f, err := parser.ParseFile(fset, "nodebuilder/"+moduleName+"/service.go", nil, parser.AllErrors|parser.ParseComments)
+		if err != nil {
+			panic(err)
+		}
+
+		cmap := ast.NewCommentMap(fset, f, f.Comments)
+
+		v := &Visitor{make(map[string]ast.Node)}
+		ast.Walk(v, f)
+
+		// TODO(@distractedm1nd): An issue with this could be two methods with the same name in different modules
+		for mn, node := range v.Methods {
+			filteredComments := cmap.Filter(node).Comments()
+			if len(filteredComments) == 0 {
+				nodeComments[mn] = "No comment exists yet for this method."
+			} else {
+				nodeComments[mn] = filteredComments[0].Text()
+			}
+		}
+	}
+	return nodeComments
+}
+
+func NewOpenRPCDocument(comments Comments) *go_openrpc_reflect.Document {
+	d := &go_openrpc_reflect.Document{}
+
+	d.WithMeta(&go_openrpc_reflect.MetaT{
+		GetServersFn: func() func(listeners []net.Listener) (*meta_schema.Servers, error) {
+			return func(listeners []net.Listener) (*meta_schema.Servers, error) {
+				return nil, nil
+			}
+		},
+		GetInfoFn: func() (info *meta_schema.InfoObject) {
+			info = &meta_schema.InfoObject{}
+			title := "Celestia Node API"
+			info.Title = (*meta_schema.InfoObjectProperties)(&title)
+
+			version := "v0.0.1"
+			info.Version = (*meta_schema.InfoObjectVersion)(&version)
+
+			description := "The Celestia Node API is the collection of RPC methods that " +
+				"can be used to interact with the services provided by Celestia Data Availability Nodes."
+			info.Description = (*meta_schema.InfoObjectDescription)(&description)
+
+			return info
+		},
+		GetExternalDocsFn: func() (exdocs *meta_schema.ExternalDocumentationObject) {
+			url := "https://github.com/celestiaorg/celestia-node/"
+			description := "Celestia Node GitHub"
```

Comment:
Can we extract those into constants somewhere above? So that we can easily find and rename those. E.g. the node repos is being renamed.

Reply:
done